runtime/lua/vim/lsp/completion.lua

```diff
@@ -501,7 +504,22 @@ end
 
 --- @param handle vim.lsp.completion.BufHandle
 local function on_insert_char_pre(handle)
+  local char = api.nvim_get_vvar('char')
+  local matched_clients = handle.triggers[char]
+
   if tonumber(vim.fn.pumvisible()) == 1 then
+    -- completion from clientA could have a trigger of clientB:
+    -- fall back to retriggering all clients (treat as incomplete)
+    if
+      matched_clients
+      and matched_clients ~= Context.current_clients
+      and vim.iter(matched_clients):any(function(client)
+        return not vim.tbl_contains(Context.current_clients, client)
+      end)
```

Comment:
I'm not happy about this piece for the corner case, but my knowledge of lua is extremely limited and it shows :)

my goto here normally, would be to track client IDs (not clients), and to use bitset for the trigger mapping and this special check (and-not).